NAAM.md

```diff
@@ -0,0 +1,278 @@
+potential confusion between the two. The new code can then be used as the Protocol ID of Naam
+Metadata.
+
+### Updating IPNS Records
+
+IPNI offers a built-in mechanism to efficiently update the metadata associated to a list of
+multihashes without having to republish the entries. Indexer nodes uniquely identify the metadata by
+the advertisement `Provider`, and `ContextID`.
+
+Both of these values remain the same in Naam advertisements: the former is the same peer ID that
+forms part of the IPNS key, i.e. `/ipns/<peer-id>`, and the latter is fixed. As a result, to update
+an IPNS record, all that's needed is to publish a new Naam advertisement that:
+
+* links to the previous advertisement in `PrevousID` link as specified by IPNI,
+* uses the same `Provider` and `ContextID` values as before,
+* sets `IsRm` to `false`, and
+* has the updated IPNS record in Naam Metadata format.
+
+Note that there is no need to specify `Entries` when updating an existing record. It is inferred by
+the network indexers since `Provider` and `ContextID` in Naam advertisements never change.
+
+### Removing IPNS Records
+
+IPNS has EOL and TTL features built-in. This means explicit removal of records is not strictly
+required. Instead, records eventually expire. Though not strictly necessary, Naam protocol can
+utilise IPNI to explicitly remove records just like a traditional DNS nameserver.
+
+The process to remove an IPNS record published via Naam protocol is identical to removal
+advertisement specified by IPNI. Similar to a regular advertisement, an IPNS record may be removed
+by publishing an advertisement that:
+
+* links to the previous advertisement in `PrevousID` link as specified by IPNI,
+* uses the same `Provider` and `ContextID` values as before,
+* sets `IsRm` to `true`, and
+* sets `Entries` link to `NoEntries` as specified by IPNI.
+
+Once ingested, indexer nodes will eventually remove the Naam advertisement metadata along with the
+publisher information.
+
+### Resolving IPNS Records
+
+IPNS record resolution involves finding the IPNS record associated to a given IPNS key, i.e.
+`/ipns/<peer-id>`. As eluded to earlier, Naam uses the find API already provided by the network
+indexers to achieve this. To resolve an IPNS record, a resolver:
+
+* calculates the multihash of the given IPNS key as its SHA-256,
+* looks up the provider info for that multihash via the IPNI find API,
+* decodes the metadata from the provider info as IPNS record,
+* validates the record, and
+* if valid returns its value.
+
+## Limitations
+
+### IPNS Record Size
+
+The maximum size of IPNS records supported by Naam is 1 KiB, even though IPNS specification suggests
+a maximum limit of 10 KiB. The 1 KiB limit currently imposed by network indexers on the maximum size
+of advertisement metadata.
+
+Basic tests show that a 1KiB limit is sufficiently large for storing simple IPNS records pointing to
```

Comment:
Some background info on the limits:

When trying to figure out a limit to use here was some of the input information https://github.com/ipfs/specs/pull/319#discussion_r970111539. The max record length found exceeded 1kib.

At the moment IPNS records have bloat to them (e.g. some data is stored in the record twice, once as cbor and another in the protobuf) so there's room to evolve the spec to cut down on used space.

You may already have metrics on this for IPNI, but my suspicion is that the current problem is not long paths but people using really large identity cids. IIRC the only way you end up needing long paths is either with non-UnixFS IPLD queries (e.g. traversing a deep path inside a single block, a deep path within an ADL that captures the query, the types of complexity that can come with the [IPLD URI scheme](https://github.com/ipfs/specs/pull/293), etc.), or with nested IPNS/dnslink queries (e.g. pointing to /ipns/otherkey/foo/bar/baz/....). Neither of these are common at the moment, people abusing identity cids definitely does happen though.

Reply:
> people using really large identity cids

Ah yes IDENTITY CIDs strike again! :) My vote would be to stop the abuse of IDENTITY CIDs, and i would go as far as saying that NAAM should reject IDENTITY CIDs ( and maybe IPFS too). I will add this as a limitation.

I am curious what is the actual usecase for putting IDENTITY CIDs inside IPNS records?

Reply:
> Ah yes IDENTITY CIDs strike again!

Yes, along with their lack of max-size specification 😭.

> I am curious what is the actual usecase for putting IDENTITY CIDs inside IPNS records?

This one's actually pretty straightforward, especially compared to standard IPNI provider records for identity CIDs.

1. Let's say Alice publishes her current favorite quote under her key with identifier `/ipns/alicekey`
2. On day 1 Alice's favorite quote is "You only live once, but if you do it right, once is enough" which is 58 characters. Rather than using SHA2-512 on it (which is 64 bytes) she decides to encode it as an identity CID to save on bytes, and uses it as the value of the IPNS record (along with the `raw` codec)
3. On day 2 Alice's favorite quote is "In the end, it's not the years in your life that count. It's the life in your years" which is 83 characters so Alice uses SHA2-512 (along with the `raw` codec) to get her CID, and updates her IPNS record to point at the CID of her new favorite quote

Basically, as long as there are any valid uses for identity CIDs, there are reasons to have them in IPNS records. Where things start getting out of hand is people pushing identity CIDs to be bigger and bigger to save themselves from doing network queries.

Reply:
Seem like forbidding identity CIDs in IPNS records would still be a reasonable restriction, or have the IPNS spec explicitly limit the length of identity CIDs to be less than the length of the largest hash.